chia/wallet/wallet_node.py

```diff
@@ -589,41 +602,55 @@ async def receive_state_from_peer(
         if fork_height is not None:
             cache.clear_after_height(fork_height)
 
-        all_tasks = []
+        all_tasks: List[asyncio.Task] = []
+        target_concurrent_tasks: int = 20
+        num_concurrent_tasks: int = 0
 
-        for idx, potential_state in enumerate(items):
+        async def receive_and_validate(inner_state: CoinState, inner_idx: int):
+            nonlocal num_concurrent_tasks
+            assert self.wallet_state_manager is not None
+            assert self.validation_semaphore is not None
+            # if height is not None:
+            async with self.validation_semaphore:
+                try:
+                    if header_hash is not None:
+                        assert height is not None
+                        self.add_state_to_race_cache(header_hash, height, inner_state)
+                        self.log.info(f"Added to race cache: {height}, {inner_state}")
+                    if trusted:
+                        valid = True
+                    else:
+                        valid = await self.validate_received_state_from_peer(inner_state, peer, cache, fork_height)
+                    if valid:
```

Comment:
`if trusted or await validate_foo():`
if you want to save the four lines

Reply:
that's harder to read and reason what's the logic without taking a sec